internal/controllers/gateway/gateway_utils.go

```diff
@@ -342,23 +343,76 @@ func getListenerStatus(
 	// https://github.com/Kong/kubernetes-ingress-controller/pull/2555#issuecomment-1154579046 for discussion)
 	// if we encountered conflicts, we must strip the ready status we originally set
 	for _, listener := range gateway.Spec.Listeners {
-		var reason string
+		var conflictReason string
+		var resolvedRefReason string
 
 		var hostname gatewayv1alpha2.Hostname
 		if listener.Hostname != nil {
 			hostname = *listener.Hostname
 		}
 		// there's no filter for protocols that don't use Hostname, but this won't be populated from earlier for those
 		if _, ok := conflictedHostnames[listener.Port][hostname]; ok {
-			reason = string(gatewayv1alpha2.ListenerReasonHostnameConflict)
+			conflictReason = string(gatewayv1alpha2.ListenerReasonHostnameConflict)
 		}
 
 		if _, ok := conflictedPorts[listener.Port]; ok {
-			reason = string(gatewayv1alpha2.ListenerReasonProtocolConflict)
+			conflictReason = string(gatewayv1alpha2.ListenerReasonProtocolConflict)
 		}
 
-		if len(reason) > 0 {
-			newConditions := []metav1.Condition{}
+		// If the listener uses TLS, we need to ensure that the gateway is granted to reference
+		// all the secrets it references
+		if listener.TLS != nil {
+			resolvedRefReason = string(gatewayv1alpha2.ListenerReasonResolvedRefs)
+			for _, certRef := range listener.TLS.CertificateRefs {
+				// only secrets are supported as certificate references
+				if (certRef.Group != nil && (*certRef.Group != "core" && *certRef.Group != "")) ||
+					(certRef.Kind != nil && *certRef.Kind != "Secret") {
+					resolvedRefReason = string(gatewayv1alpha2.ListenerReasonInvalidCertificateRef)
+					break
+				}
+				// if the certificate is in the same namespace of the gateway, no ReferenceGrant is needed
+				if certRef.Namespace == nil {
```

Comment:
Technically certificate refs could have a ref with a namespace that is the same as the `Gateway` but that namespace explicitly called out in the listener, right? Even if we don't expect this with our current controller, it would be valid behavior in the future so we might as well add the conditional here:

```suggestion
				if certRef.Namespace == nil || *certRef.Namespace == gatewayv1alpha2.Namespace(gateway.Namespace) {
```

Is this right or am I missing something? :thinking:

Reply:
yes, you are right! I'm going to add this condition